internal/framework/status/setters.go

```diff
@@ -0,0 +1,215 @@
+)
+
+// setter is a function that takes an object and sets the status on that object if the status has changed.
+// If the status has not changed, and the setter does not set the status, it returns false.
+type setter func(client.Object) bool
+
+func newNginxGatewayStatusSetter(clock Clock, status NginxGatewayStatus) func(client.Object) bool {
+	return func(object client.Object) bool {
+		ng := object.(*ngfAPI.NginxGateway)
+		conds := convertConditions(
+			status.Conditions,
+			status.ObservedGeneration,
+			clock.Now(),
+		)
+
+		if conditionsEqual(ng.Status.Conditions, conds) {
+			return false
+		}
+
+		ng.Status = ngfAPI.NginxGatewayStatus{
+			Conditions: conds,
+		}
+
+		return true
+	}
+}
+
+func newGatewayClassStatusSetter(clock Clock, gcs GatewayClassStatus) func(client.Object) bool {
+	return func(object client.Object) bool {
+		gc := object.(*v1beta1.GatewayClass)
+		status := prepareGatewayClassStatus(gcs, clock.Now())
+
+		if conditionsEqual(gc.Status.Conditions, status.Conditions) {
+			return false
+		}
+
+		gc.Status = status
+
+		return true
+	}
+}
+
+func newGatewayStatusSetter(podIP string, clock Clock, gs GatewayStatus) func(client.Object) bool {
+	return func(object client.Object) bool {
+		gw := object.(*v1beta1.Gateway)
+		status := prepareGatewayStatus(gs, podIP, clock.Now())
+
+		if gwStatusEqual(gw.Status, status) {
+			return false
+		}
+
+		gw.Status = status
+
+		return true
+	}
+}
+
+func newHTTPRouteStatusSetter(gatewayCtlrName string, clock Clock, rs HTTPRouteStatus) func(client.Object) bool {
+	return func(object client.Object) bool {
+		hr := object.(*v1beta1.HTTPRoute)
+		// statuses.GatewayStatus is never nil when len(statuses.HTTPRouteStatuses) > 0
+		status := prepareHTTPRouteStatus(
+			rs,
+			gatewayCtlrName,
+			clock.Now(),
+		)
+
+		if hrStatusEqual(gatewayCtlrName, hr.Status, status) {
+			return false
+		}
+
+		hr.Status = status
+
+		return true
+	}
+}
+
+func gwStatusEqual(prev, cur v1beta1.GatewayStatus) bool {
+	addressesEqual := slices.EqualFunc(prev.Addresses, cur.Addresses, func(a1, a2 v1beta1.GatewayStatusAddress) bool {
+		if deref[v1beta1.AddressType](a1.Type) != deref[v1beta1.AddressType](a2.Type) {
+			return false
+		}
+
+		return a1.Value == a2.Value
+	})
+
+	if !addressesEqual {
+		return false
+	}
+
+	if !conditionsEqual(prev.Conditions, cur.Conditions) {
+		return false
+	}
+
+	return slices.EqualFunc(prev.Listeners, cur.Listeners, func(s1, s2 v1beta1.ListenerStatus) bool {
+		if s1.Name != s2.Name {
+			return false
+		}
+
+		if s1.AttachedRoutes != s2.AttachedRoutes {
+			return false
+		}
+
+		if !conditionsEqual(s1.Conditions, s2.Conditions) {
+			return false
+		}
+
+		return slices.EqualFunc(s1.SupportedKinds, s2.SupportedKinds, func(k1, k2 v1beta1.RouteGroupKind) bool {
+			if k1.Kind != k2.Kind {
+				return false
+			}
+
+			return deref(k1.Group) == deref(k2.Group)
+		})
+	})
+}
+
+func hrStatusEqual(gatewayCtlrName string, prev, cur v1beta1.HTTPRouteStatus) bool {
```

Comment:
I got a bit confused while reading this function.

I think there are two primary cases here:
- First, mismatch in parents (parentRefs) -- catch cases when curr doesn't have old parents or curr has completely new parents.
- Second, when parents exist in both prev and curr, ensure they conditions match.

However, the comments talk only about conditions.

I wonder if the following implementation is simpler -- I tried to make the two cases above explicit:

```go
func hrStatusEqual(gatewayCtlrName string, prev, cur v1beta1.HTTPRouteStatus) bool {
	// Since other controllers may update HTTPRoute status we can't assume anything about the order of the statuses,
	// and we have to ignore statuses written by other controllers when checking for equality.
	// Therefore, we can't use slices.EqualFunc here because it cares about the order.

	// First check if prev status has any parents that are no longer present in the cur status.
	for _, prevParent := range prev.Parents {
		if prevParent.ControllerName != v1beta1.GatewayController(gatewayCtlrName) {
			continue
		}

		idx := slices.IndexFunc(cur.Parents, func(curParent v1beta1.RouteParentStatus) bool {
			return parentStatusEqualParentRef(prevParent, curParent)
		})

		if idx == -1 {
			return false
		}
	}

	// Then, we check if cur status has any parents that are no longer present in the prev status.
	// Also, if parents are present in both statuses, we check if their conditions are equal.
	for _, curParent := range cur.Parents {
		idx := slices.IndexFunc(prev.Parents, func(prevParent v1beta1.RouteParentStatus) bool {
			return parentStatusEqualParentRef(curParent, prevParent)
		})

		if idx == -1 {
			return false
		}

		if !conditionsEqual(curParent.Conditions, prev.Parents[idx].Conditions) {
			return false
		}
	}

	return true
}
. . .

func parentStatusEqualParentRef(p1, p2 v1beta1.RouteParentStatus) bool {
	if p1.ControllerName != p2.ControllerName {
		return false
	}

	if p1.ParentRef.Name != p2.ParentRef.Name {
		return false
	}

	if deref(p1.ParentRef.Namespace) != deref(p2.ParentRef.Namespace) {
		return false
	}

	if deref(p1.ParentRef.SectionName) != deref(p2.ParentRef.SectionName) {
		return false
	}

	return true
}
```

Reply:
This misses the case where the current parentRef exists in the previous status, but the conditions differ. The conditions belong to the parentRef, so I think it makes sense to check the conditions in the `routeParentStatusEqual` function.

Reply:
> This misses the case where the current parentRef exists in the previous status, but the conditions differ. 

```
		if !conditionsEqual(curParent.Conditions, prev.Parents[idx].Conditions) {
			return false
		}
```
that covers it 🤔

Reply:
In the first for loop that is missing

Reply:
I changed the comments to clarify what's happening in the function, but I don't see the advantage of moving the condition check out of the `routeParentStatusEqual` function. I personally think it makes more sense in the function since the conditions are a part of the RouteParentStatus.

Reply:
> I changed the comments to clarify what's happening in the function, but I don't see the advantage of moving the condition check out of the routeParentStatusEqual function. I personally think it makes more sense in the function since the conditions are a part of the RouteParentStatus.

sounds good 👍 

> In the first for loop that is missing

well, you only need to do it one loop, because both loops find the intersection of parentRefs